tests/test_main.py

```diff
@@ -0,0 +1,371 @@
+
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 1
+
+        crud_event = crud_event_qs.first()
+        assert crud_event.user is None
+
+    def test_middleware_logged_in_user_in_request(self, user, client):
+        client.force_login(user)
+        create_obj_url = reverse("test_app:create-obj")
+        client.post(create_obj_url)
+        assert Model.objects.count() == 1
+
+        obj = Model.objects.first()
+        crud_event = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        ).first()
+        assert crud_event.user == user
+
+
+@pytest.mark.asyncio
+@pytest.mark.django_db(transaction=True)
+class TestASGIRequestEvent:
+    async def test_login(self, async_user, async_client, username, password):
+        await sync_to_async(async_client.login)(username=username, password=password)
+        assert await sync_to_async(RequestEvent.objects.count)() == 0
+
+        resp = await async_client.get(reverse("test_app:index"))
+        assert resp.status_code == 200
+
+        qs = await sync_to_async(RequestEvent.objects.filter)(user=async_user)
+        assert await sync_to_async(qs.exists)()
+
+    async def test_remote_addr_default(self, async_client):
+        assert await sync_to_async(RequestEvent.objects.count)() == 0
+
+        resp = await async_client.request(
+            method="GET",
+            path=str(reverse("test_app:index")),
+            server=("127.0.0.1", "80"),
+            scheme="http",
+            headers=[(b"host", b"testserver")],
+            query_string="",
+        )
+        assert resp.status_code == 200
+
+        event = await sync_to_async(RequestEvent.objects.get)(url=reverse("test_app:index"))
+        assert event.remote_ip == "127.0.0.1"
+
+    async def test_remote_addr_another(self, async_client):
+        assert await sync_to_async(RequestEvent.objects.count)() == 0
+
+        resp = await async_client.request(
+            method="GET",
+            path=str(reverse("test_app:index")),
+            server=("127.0.0.1", "80"),
+            client=("10.0.0.1", 111),
+            scheme="http",
+            headers=[(b"host", b"testserver")],
+            query_string="",
+        )
+        assert resp.status_code == 200
+
+        event = await sync_to_async(RequestEvent.objects.get)(url=reverse("test_app:index"))
+        assert event.remote_ip == "10.0.0.1"
+
+
+@pytest.mark.django_db
+class TestWSGIRequestEvent:
+    def test_login(self, user, client, username, password):
+        client.login(username=username, password=password)
+        assert RequestEvent.objects.count() == 0
+
+        resp = client.get(reverse("test_app:index"))
+        assert resp.status_code == 200
+
+        assert RequestEvent.objects.get(user=user)
+
+
+@pytest.mark.django_db
+class TestAuditAdmin:
```

Comment:
Changes to this test removed the need for `BeautifulSoup`, it now uses Django's built-in `assertInHTML`. Also of note is that the HTML structure changed in Django 4.1 to use `summary` tags instead of `h3`